### ISSUE TITLE
Support JSONPath on JsonDocument

### ISSUE THREAD

**Issue by steveharter** [dotnet/corefx#38074]
As an extension to https://github.com/dotnet/corefx/issues/37768, the JsonDocument class should support setting the `Path` property on `JsonReaderException`.

In addition, once this is complete, the serializer should re-throw `JsonReaderException` (as it does today) but now with the `Path` property pre-pended with the results from `JsonReaderException` to include the full Path of the serializer + JsonDocument.

See also PR https://github.com/dotnet/corefx/pull/37938

**Comment by MarcoRossignoli** [comment]
@steveharter is this fixed?Should we close?

**Comment by ahsonkhan** [comment]
cc @bartonjs

**Comment by bartonjs** [comment]
Is the idea that the database building routine be in a try/catch where the catch modifies the exception object and then continues the same callstack/dispatch with `throw;`?  Next most obvious alternative: That in the catch it throws a new Exception, wrapping the old one, and the outer one has the path set.

Either of those seems doable; either by making the StackRow value also keep track of the StartObject/StartArray (which would also eliminate the seek-backwards for every EndObject/EndArray at the expense of needing to grow the temporary buffer more often) or changing/overloading `FindIndexOfFirstUnsetSizeOrLength` to accept a "no later than" parameter. -- The StackRowStack knows how many elements already exist in the array (maybe off-by-one, but the data is there), and for an object property the name is available from the node after StartArray.

It's probably something easy enough for up-for-grabs.

I'd do it by renaming 
```C#
        private static void Parse(
            ReadOnlySpan<byte> utf8JsonSpan,
            Utf8JsonReader reader,
            ref MetadataDb database,
            ref StackRowStack stack)
        {
```

to `ParseCore`, and making a new Parse with the same signature which just does

```C#
private static void Parse(
    ReadOnlySpan<byte> utf8JsonSpan,
    Utf8JsonReader reader,
    ref MetadataDb database,
    ref StackRowStack stack)
{
    try
    {
        ParseCore(utf8JsonSpan, reader, ref database, ref stack);
    }
    catch (JsonReaderException e)
    {
        // Using stack, database, and maybe Stack<string> parts, build the path
        e.Path = path;
    }
}
```

**Comment by gary-holland** [comment]
Hi,

Is JsonPath support for querying the JsonDocument/JsonElement structures still being looked at, as this is a major gap for us shifting from Newtonsoft to system.text.json.

What we need is an equivalent to:
```csharp
var jsonPath = "$.my.path";
var json = JToken.Parse(jsonString);
var token = json.SelectToken(jsonPath);
```

Something like this would be ideal:
```csharp
var jsonPath = "$.my.path";
var jsonDoc = JsonDocument.Parse(json);
var element = jsonDoc.SelectElement(jsonPath); //returns JsonElement
var elements = jsonDoc.SelectElements(jsonPath); //returns JsonElement.ArrayEnumerator
```

Thanks.

**Comment by ahsonkhan** [comment]
> Is JsonPath support for querying the JsonDocument/JsonElement structures still being looked at

This issue is specifically about storing the JSON path information as part of the exception whenever the JsonDocument sees an invalid JSON while parsing, and not about full-fledged JsonPath support for querying the JSON data within the JsonDocument. Since this would be a separate feature, please create a separate issue for it. I would imagine the path support (at least for your use case), primarily be for convenience (requiring less code).

**Comment by gary-holland** [comment]
Thanks. @ahsonkhan.  I've created a new issue  #41537
